src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLiteConnectionPool.cs

```diff
@@ -0,0 +1,151 @@
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.SQLite.v2.Interop;
+
+namespace Microsoft.CodeAnalysis.SQLite.v2
+{
+    internal sealed partial class SQLiteConnectionPool : IDisposable
+    {
+        // We pool connections to the DB so that we don't have to take the hit of 
+        // reconnecting.  The connections also cache the prepared statements used
+        // to get/set data from the db.  A connection is safe to use by one thread
+        // at a time, but is not safe for simultaneous use by multiple threads.
+        private readonly object _connectionGate = new();
+        private readonly Stack<SqlConnection> _connectionsPool = new();
+
+        private readonly CancellationTokenSource _shutdownTokenSource = new();
+
+        private readonly SQLiteConnectionPoolService _connectionPoolService;
+        private readonly IPersistentStorageFaultInjector? _faultInjector;
+        private readonly string _databasePath;
+        private readonly IDisposable _ownershipLock;
+
+        public SQLiteConnectionPool(SQLiteConnectionPoolService connectionPoolService, IPersistentStorageFaultInjector? faultInjector, string databasePath, IDisposable ownershipLock)
+        {
+            _connectionPoolService = connectionPoolService;
+            _faultInjector = faultInjector;
+            _databasePath = databasePath;
+            _ownershipLock = ownershipLock;
+        }
+
+        internal void Initialize(
+            Solution? bulkLoadSnapshot,
+            Action<Solution?, SqlConnection, CancellationToken> initializer,
+            CancellationToken cancellationToken)
+        {
+            // This is our startup path.  No other code can be running.  So it's safe for us to access a connection that
+            // can talk to the db without having to be on the reader/writer scheduler queue.
+            using var _ = GetPooledConnection(checkScheduler: false, out var connection);
+
+            initializer(bulkLoadSnapshot, connection, cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            // Flush all pending writes so that all data our features wanted written
+            // are definitely persisted to the DB.
+            try
+            {
+                _shutdownTokenSource.Cancel();
+                CloseWorker();
+            }
+            finally
+            {
+                // let the lock go
+                _ownershipLock.Dispose();
+            }
+        }
+
+        private void CloseWorker()
+        {
+            lock (_connectionGate)
+            {
+                // Go through all our pooled connections and close them.
+                while (_connectionsPool.Count > 0)
+                {
+                    var connection = _connectionsPool.Pop();
+                    connection.Close_OnlyForUseBySqlPersistentStorage();
```

Comment:
this method name may need to change.  i.e. Close_OnlyForUseBySQLiteConnectionPool

Reply:
➡️ Renamed